src/lambda_toolshed/papillon.cljc

```diff
@@ -24,150 +15,169 @@
+  process the interceptor chain `ixs`."
   [ctx ixs]
   (assoc (enqueue ctx ixs)
-         :lambda-toolshed.papillon/stack []))
+         ::stack []))
+
+(defn- present-sync
+  [result]
+  (if-let [error (::error result)]
+    (throw error)
+    result))
 
-(defn- await-result
-  "'Unwinds' any nested channels and returns the context"
-  [res]
-  (go-loop [ctx res]
+(defn- present-async
+  [result]
+  (go-loop [ctx result]
     (if (satisfies? ReadPort ctx)
       (recur (<! ctx))
-      ctx)))
+      (if-let [error (::error ctx)]
+        error
+        ctx))))
+
+(defn- namer [i ix]
+  (if (:name ix)
+    ix
+    (let [fmt #?(:clj format :cljs gstring/format)]
+      (vary-meta ix update :name (fnil identity (fmt "itx%02d" i))))))
 
 (defn execute
-  "Executes the interceptor call chain as a queue.
-
-  It will run foward through the chain calling the function
-  associated to the `:enter` key where that where that function
-  exists, while adding the interceptor to the history of
-  interceptors seen, so when the :enter chain is
-  finished, it can run backwards through the history
-  (reverse order) to apply the functions associated to `:leave`,
-  or `:error`, as determined by the `:lambda-toolshed.papillon/error` key
-  on the context.
-
-  `execute` takes any map as an initial context, and will associate
-  the queue and stack into the context to start processing.  If no
-  context is provided an empty map is used.  Note: The behavior of
-  starting with an initial context that contains the key
-  `:lambda-toolshed.papillon/error` is left unspecified and may be subject to
-  change."
+  "Execute the interceptor chain `ixs` with the given initial context `ctx`.
+
+  Run foward through the chain calling the functions associated with the
+  `:enter` key (where it exists), while accumulating a stack of interceptors
+  seen.
+
+  When the `:enter` chain is exhausted, run the accumulated stack in reverse
+  order invoking the function at the `:leave` or `:error` key based on the
+  `:lambda-toolshed.papillon/error` key of the context.
+
+  The initial context `ctx` is augmented with the necessary housekeeping data
+  structures before processing the chain.  If no context is provided an empty
+  map is used.
+
+  Note: Starting with an initial context that contains the key
+  `:lambda-toolshed.papillon/error` is undefined and may change.
+
+  Executing the interceptor chain can complete synchronously or asynchronously.
+  If any interceptor function (enter, leave or error) completes asynchronously,
+  then chain execution will complete asynchronously, otherwise the chain execution
+  will complete synchronously.
+
+  The result of executing the chain is either:
+
+  1. (async) a channel whose sole value is the chain execution result or an
+     exception (when not handled by the chain).
+  2. (sync) the chain execution result, or a thrown exception (when not handled
+     by the chain)."
   ([ixs]
    (execute {} ixs))
```

Comment:
This needs to have the order swapped here.

Reply:
Done.